Manually merge main -> rebranch

bnbarham opened the pull request:
Conflicts:
  - `lib/ClangImporter/ClangModuleDependencyScanner.cpp` not actually
    sure, don't see any changes in `bridgeClangModuleDependencies` that
    are rebranch specific
  - `utils/build.ps1` clang-psuedo-gen was removed on rebranch

compnerd committed:
utils: build the dynamic experimental SDK

slavapestov committed:
One final round of small cleanups for Monoids benchmark

Commit:
Reapply [cxx-interop] Avoid copies when accessing pointee

After #83289 and #82879 landed we should no longer get deserialization
failures and this feature is no longer behind a flag. This patch also
changes how we query if a function's return value depends on self.
Previously, we queried the lifetime dependencies from the Swift
declaration. Unfortunately, this is problematic as we might have not
finished fully importing the types in the function signature just yet
and the compiler might end up populating the conformance tables
prematurely. To work this around, I store functions with self-dependent
return values where lifetimes are computed in the importer for later
use.

The PR also adds a test to make sure the addressable dependency feature
will not result in deserialization errors.

rdar://155319311&154213694&112690482&128293252

artemcm committed:
[Dependency Scanning] Bridge Clang dependency scanner results on-demand

Instead of always bridging all of the discovered modules of all of the queries, only do so for modules which are not already cached

artemcm committed:
[Dependency Scanning] Refactor Clang dependency bridging into a 'ModuleDependencyScanner' utility

This moves the functionality of 'bridgeClangModuleDependency' into a utility in the main scanner class because it relies on various objects whose lifetime is already tied to the scanner itself.

MaxDesiatov committed:
Build WasmKit with host SwiftPM when SwiftPM is not built

Currently, passing `--wasmkit` to `build-script` also requires `--swiftpm --install-swiftpm` and all of their dependencies (at the very least Dispatch, Foundation, XCTest, Testing, llbuild, swift-build). We can conditionally switch to host SwiftPM in this case.

MaxDesiatov committed:
Update utils/swift_build_support/swift_build_support/products/wasmkit.py

compnerd committed:
build.ps1: improve documentation (NFC)

Update the documentation for the parameters, remove the public
documentation for "advanced" options (e.g. to control migration, CI
schemes, etc).

compnerd committed:
build.ps1: intrduce the ability to build with clang optionally

Add a `-UseHostCompiler` option which allows you to to pass in
`-UseHostCompiler=$False` to switch to using the bootstrap compiler to
build the toolchain.

compnerd committed:
build.ps1: attempt to repair the ARM64 cross-compile

MaxDesiatov committed:
Make logging in `wasmkit.py` resilient to file renaming

Commit:
[cxx-interop] Make reverse interop header compile in embedded mode

These are the minimal changes to make the reverse interop header compile
when embedded Swift is used. Previously, the _SwiftStdlibCxxOverlay.h
and the generated interop header disagreed on the mangling of a symbol.

This is not sufficient yet to use the embedded Swift standard library from
reverse interop, there are some missing symbols while linking. But this
PR enables doing reverse interop without using the stdlib types like
swift::String in C++. Follow-up PRs will fix the rest of the issues.

rdar://154740225

Commit:
[cxx-interop] Make sure we emit lifetime warnings for templated classes

rdar://155542619

egorzhdan committed:
[cxx-interop] Prohibit weak references to foreign reference types

This fixes a runtime crash when a `weak` reference to a C++ foreign reference type is used.

Instead of a runtime crash, Swift would now emit a compiler error saying that `weak` keyword is incompatible with foreign reference types.

rdar://124040825 / resolves https://github.com/swiftlang/swift/issues/83080

Xazax-hun committed:
Merge pull request #83370 from Xazax-hun/addressable-param-copy-reapply

Reapply [cxx-interop] Avoid copies when accessing pointee

artemcm committed:
Merge pull request #83600 from artemcm/NoRedundantClangDepBridging

[Dependency Scanning] Bridge Clang dependency scanner results on-demand

MaxDesiatov committed:
Merge pull request #83755 from MaxDesiatov/wasmkit-cmake

Currently, passing `--wasmkit` to `build-script` also requires `--swiftpm --install-swiftpm` and all of their dependencies (at the very least Dispatch, Foundation, XCTest, Testing, llbuild, swift-build). We can conditionally switch to host SwiftPM when new SwiftPM is not built or installed.

Xazax-hun committed:
Merge pull request #83785 from Xazax-hun/embedded-swift-reverse-interop

[cxx-interop] Make reverse interop header compile in embedded mode

kubamracek committed:
[ClangImporter] Skip importing values for ObjCBool declarations

aschwaighofer committed:
Merge pull request #83783 from compnerd/arm64

build.ps1: attempt to repair the ARM64 cross-compile

MaxDesiatov committed:
Embedded Wasm: enable `import WASILibc`

WASILibc wasn't built for the embedded stdlib in `stdlib/public/Platform/CMakeLists.txt`. New `copy_wasilibc_modulemap_embedded_resource` and `embedded-stdlib-platform-${triple}` targets are added, the latter for `wasm32-unknown-wasip1` only for now. Also added a `wasilibc_functions.swift` test to verify the result.

Resolves https://github.com/swiftlang/swift/issues/83513

rdar://157467412

compnerd committed:
Merge pull request #83781 from compnerd/documentation

egorzhdan committed:
[cxx-interop] Validate C foreign reference types

Swift validates the retain/release operations for foreign reference types to check for obvious errors, e.g. a wrong parameter type or return type.

That logic was only running for C++ foreign reference types. This patch enables it for C foreign reference types as well.

rdar://158609723

drexin committed:
[Test] Disable Interpreter/typed_throws_abi.swift on back deployment runtime

rdar://158442662

This is not supported on back deployment runtimes.

Xazax-hun committed:
Merge pull request #83786 from Xazax-hun/diag-for-templated-classes

[cxx-interop] Make sure we emit lifetime warnings for templated classes

kubamracek committed:
[embedded] When loading (sub-)modules, propagate the Embedded flag and use it for hash computation

MaxDesiatov committed:
Platform/CMake: add `embedded-stdlib-${mod}` dependency

MaxDesiatov committed:
Platform/CMake: remove unused `message` func, `mod` var

tshortli committed:
AST: Allow obsoletion in Swift version to disambiguate overloads consistently.

It must be possible to disambiguate overloads using the module-wide Swift
language version, even in contexts that are themselves obsolete in the current
Swift language version.

Resolves rdar://158620835.

kubamracek committed:
[embedded] Also specify -enable-experimental-feature Embedded to sub-invocations

slavapestov committed:
Merge pull request #83797 from slavapestov/tweak-monoids-benchmark

One final round of small cleanups for Monoids benchmark

tshortli committed:
Merge pull request #83799 from tshortli/obsolete-in-version-of-swift-ambiguity

AST: Allow obsoletion in Swift version to disambiguate overloads consistently

drexin committed:
Merge pull request #83794 from drexin/wip-158442662

[Test] Disable Interpreter/typed_throws_abi.swift on back deployment …

MaxDesiatov committed:
Platform/CMake: use `arch` instead of `arch_subdir`

This fixes wasi-libc.modulemap not found after installation

egorzhdan committed:
Merge pull request #83787 from egorzhdan/egorzhdan/weak-reference

[cxx-interop] Prohibit weak references to foreign reference types

egorzhdan committed:
Merge pull request #83793 from egorzhdan/egorzhdan/c-frt-validation

[cxx-interop] Validate C foreign reference types

egorzhdan committed:
[cxx-interop] Avoid ambiguous C++ overloads for functions from Swift extensions

Reverse interop has logic that avoids emitting ambiguous overloads into the generated header. That logic did not apply for functions declared within Swift extensions. This meant that the generated C++ header would sometimes not compile.

rdar://158252800

kubamracek committed:
Merge pull request #83791 from kubamracek/const-objcbool

[ClangImporter] Skip importing values for ObjCBool declarations

kubamracek committed:
Merge pull request #83795 from kubamracek/embedded-propagate-to-submodule-builds

[embedded] When loading (sub-)modules, propagate the Embedded flag and use it for hash computation

compnerd committed:
Merge pull request #83717 from compnerd/dynamic-experimental-sdk

utils: build the dynamic experimental SDK

compnerd committed:
Merge pull request #83796 from compnerd/clang

build.ps1: intrduce the ability to build with clang optionally

MaxDesiatov committed:
Embedded Wasm: enable `import WASILibc` (#83792)

WASILibc wasn't built for the embedded stdlib in `stdlib/public/Platform/CMakeLists.txt`. New `copy_wasilibc_modulemap_embedded_resource` and `embedded-stdlib-platform-${triple}` targets are added, the latter for `wasm32-unknown-wasip1` only for now. Also added a `wasilibc_functions.swift` test to verify the result.

Resolves https://github.com/swiftlang/swift/issues/83513

rdar://157467412

egorzhdan committed:
Merge pull request #83811 from egorzhdan/egorzhdan/ambiguous-init

[cxx-interop] Avoid ambiguous C++ overloads for functions from Swift extensions

bnbarham commented:
@swift-ci please test

AnthonyLatsis commented:
> don't see any changes in `bridgeClangModuleDependencies` that are rebranch specific

https://github.com/swiftlang/swift/commit/5e09125

artemcm reviewed (approved):
`lib/ClangImporter/ClangModuleDependencyScanner.cpp` looks fine, I also don't see anything Rebranch specific here.

AnthonyLatsis committed:
Mnually merge main -> rebranch

Conflicts:
* `lib/ClangImporter/ClangModuleDependencyScanner.cpp`:
  https://github.com/swiftlang/swift/commit/5e09125
*	`utils/build.ps1`: https://github.com/swiftlang/swift/pull/83523

bnbarham commented:
Sigh, I don't know how I missed that line :(. Thanks @AnthonyLatsis!

bnbarham commented:
Up to tests, merging.
